python/cugraph-nx/cugraph_nx/classes/graph.py

```diff
@@ -0,0 +1,133 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import annotations
+
+from collections.abc import Hashable, Iterator
+from typing import TYPE_CHECKING
+
+import cugraph_nx as cnx
+
+if TYPE_CHECKING:
+    import cupy as cp
+
+NodeType = Hashable
+
+
+__all__ = ["Graph"]
+
+
+class Graph:
+    indptr: cp.ndarray
+    row_indices: cp.ndarray
+    col_indices: cp.ndarray
+    edge_values: cp.ndarray | None
+    key_to_id: dict[NodeType, int] | None
+    _id_to_key: dict[int, NodeType] | None
+    _N: int
+    graph: dict
+
+    __networkx_plugin__ = "cugraph"
+
+    graph_attr_dict_factory = dict
+
+    def __init__(
+        self,
+        indptr,
+        row_indices,
+        col_indices,
+        edge_values,
+        *,
+        key_to_id,
+        id_to_key=None,
+        **attr,
+    ):
+        # TODO: when should we make and wrap a plc.Graph? Can it own these arrays?
+        self.indptr = indptr
+        self.row_indices = row_indices
+        self.col_indices = col_indices
+        self.edge_values = edge_values
+        self.key_to_id = key_to_id
+        self._id_to_key = id_to_key
+        self._N = indptr.size - 1
+        self.graph = self.graph_attr_dict_factory()
+        self.graph.update(attr)
+
+    @property
+    def id_to_key(self) -> dict[int, NodeType]:
+        if self.key_to_id is None:
+            return None
+        if self._id_to_key is None:
+            self._id_to_key = {val: key for key, val in self._key_to_id.items()}
+        return self._id_to_key
+
+    def __iter__(self) -> Iterator:
+        if self.key_to_id is not None:
+            return iter(self.key_to_id)
+        return iter(range(self._N))
+
+    def __contains__(self, n) -> bool:
+        if self.key_to_id is not None:
+            container = self.key_to_id
+        else:
+            container = range(self._N)
+        try:
+            return n in container
+        except TypeError:
+            return False
+
```

Comment:
how many of these function are truly used within an algorithms?  There is value to those functions, but it seems like Nx could just leverage their internal representation first for those calls.   

Now if the thought that Nx would totally off load all functions and just be a broker?

Reply:
I think methods on graphs are "nice to have", but most of them aren't strictly necessary for the example workloads we're after. I only added the necessary ones and the "easy to implement" ones. Long term, having methods on graphs will let users copy/paste examples from documentation, docstrings, etc. and update to use cugraph-nx. It gives users a familiar experience that may be useful when interacting with graphs in a notebook. I don't expect to expend significant effort in the near term to implement the missing methods.